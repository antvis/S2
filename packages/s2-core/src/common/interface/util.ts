export type DeepRequired<T extends Record<string, any>> = {
  [K in keyof T]-?: NonNullable<T[K]> extends Record<string, any>
    ? DeepRequired<NonNullable<T[K]>>
    : NonNullable<T[K]>;
};
