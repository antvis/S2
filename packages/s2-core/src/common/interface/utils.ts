export type PickEssential<O> = {
  [K in keyof O as Pick<Partial<O>, K> extends Pick<O, K> ? never : K]: O[K];
};
