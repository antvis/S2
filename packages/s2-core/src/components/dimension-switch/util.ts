import { DimensionType, DimensionItem } from './dimension';

export interface OperatedType {
  [type: string]: string[];
}

export const getDimensionsByPredicate = (
  data: DimensionType[],
  predicate: (value: DimensionItem) => boolean,
) => {
  return data.reduce((obj, t) => {
    obj[t.type] = t.items.filter(predicate).map((i) => i.id);
    return obj;
  }, {} as OperatedType);
};
