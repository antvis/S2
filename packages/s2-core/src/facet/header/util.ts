import { isNil } from 'lodash';

export const getCellPadding = () => {
  const padding = [12, 4, 12, 4];
  const left = isNil(padding[3]) ? 0 : padding[3];
  const right = isNil(padding[1]) ? 0 : padding[1];
  const top = isNil(padding[0]) ? 0 : padding[0];
  const bottom = isNil(padding[2]) ? 0 : padding[2];
  return {
    left,
    right,
    top,
    bottom,
  };
};
