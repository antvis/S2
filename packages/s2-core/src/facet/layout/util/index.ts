export { default as getDimsConditionByNode } from './get-dims-condition-by-node';
export { default as getColHeight } from './get-col-height';
export { default as getColWidth } from './get-col-width';
export { default as getRowWidth } from './get-row-width';
export { default as processColsList } from './process-cols-list';
export { default as getHeaderHierarchyQuery } from './get-header-hierarchy-query';
