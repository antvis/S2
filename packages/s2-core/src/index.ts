export { Event as GEvent } from '@antv/g-canvas';
export { getTheme } from './theme';
export { Node } from './facet/layout/node';
export { Hierarchy } from './facet/layout/hierarchy';
export { buildTableHierarchy } from './facet/layout/build-table-hierarchy';
export * from './common';
export * from './utils';
export * from './cell';
export * from './sheet-type';
export * from './data-set';
export * from './interaction';
export * from './ui/tooltip';
export * from './facet/header';
