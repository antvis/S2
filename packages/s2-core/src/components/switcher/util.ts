import { filter, flatten, isEmpty, isNil, map } from 'lodash';
import { DraggableLocation } from 'react-beautiful-dnd';
import {
  FieldType,
  MAX_DIMENSION_COUNT,
  SWITCHER_PREFIX_CLS,
} from './constant';
import { SwitcherItem, SwitchResult, SwitchState } from './interface';
import { getClassNameWithPrefix } from '@/utils/get-classnames';

export const getSwitcherClassName = (...classNames: string[]) =>
  getClassNameWithPrefix(SWITCHER_PREFIX_CLS, ...classNames);

export const getNonEmptyFieldCount = (state: SwitchState) => {
  return [
    state[FieldType.Rows],
    state[FieldType.Cols],
    state[FieldType.Values],
  ].reduce((sum, value) => sum + (isEmpty(value) ? 0 : 1), 0);
};

export const getMainLayoutClassName = (nonEmptyCount: number) => {
  switch (nonEmptyCount) {
    case 1:
      return getSwitcherClassName('content', 'one-dimension');
    case 2:
      return getSwitcherClassName('content', 'two-dimensions');
    default:
      return getSwitcherClassName('content', 'three-dimensions');
  }
};

export const showDimensionCrossRows = (nonEmptyCount: number) =>
  nonEmptyCount < MAX_DIMENSION_COUNT;

export const isMeasureType = (fieldType: FieldType) =>
  fieldType === FieldType.Values;

export const moveItem = (
  source: SwitcherItem[],
  destination: SwitcherItem[],
  droppableSource: DraggableLocation,
  droppableDestination: DraggableLocation,
): SwitchState => {
  // change order in same column
  if (droppableDestination.droppableId === droppableSource.droppableId) {
    const updatingDestination = [...destination];
    const [removed] = updatingDestination.splice(droppableSource.index, 1);
    updatingDestination.splice(droppableDestination.index, 0, removed);
    return {
      [droppableDestination.droppableId]: updatingDestination,
    };
  }
  // move to other column
  const updatingSource = [...source];
  const updatingDestination = [...destination];

  const [removed] = updatingSource.splice(droppableSource.index, 1);
  updatingDestination.splice(droppableDestination.index, 0, removed);

  return {
    [droppableSource.droppableId]: updatingSource,
    [droppableDestination.droppableId]: updatingDestination,
  };
};

export const checkItem = (
  source: SwitcherItem[],
  checked: boolean,
  id: string,
  parentId?: string,
): SwitcherItem[] => {
  const target: SwitcherItem = {
    ...source.find((item) => item.id === (parentId ?? id)),
  };

  // 有 parentId 时，说明是第二层次项的改变
  if (parentId) {
    target.children = map(target.children, (item) => ({
      ...item,
      checked: item.id === id ? checked : item.checked,
    }));
  } else {
    target.checked = checked;
    target.children = map(target.children, (item) => ({
      ...item,
      checked: checked,
    }));
  }

  return source.map((item) => (item.id === target.id ? target : item));
};

export const generateSwitchResult = (state: SwitchState): SwitchResult => {
  // rows and cols can't be hidden
  const rows = map(state[FieldType.Rows], 'id');
  const cols = map(state[FieldType.Cols], 'id');

  // flatten all values and derived values
  const values = flatten(
    map(state[FieldType.Values], (item) => {
      const children = map(item.children, 'id');
      return [item.id, ...children];
    }),
  );

  const filterHiddenValues = (item: SwitcherItem) => item.checked === false;

  //  get all hidden values
  const hiddenValues = flatten(
    map(filter(state[FieldType.Values], filterHiddenValues), (item) => {
      const hiddenChildren = map(
        filter(item.children, filterHiddenValues),
        'id',
      );
      return [item.id, ...hiddenChildren];
    }),
  );

  return { rows, cols, values, hiddenValues };
};
