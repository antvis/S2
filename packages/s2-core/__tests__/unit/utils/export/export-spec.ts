import type { S2DataConfig, SpreadSheet } from '../../../../src';
import { asyncGetAllPlainData } from '../../../../src/utils';
import { customRowGridSimpleFields } from '../../../data/custom-grid-simple-fields';
import { CustomGridData } from '../../../data/data-custom-grid';
import { assembleDataCfg, assembleOptions } from '../../../util';
import { getContainer } from '../../../util/helpers';
import { PivotSheet, TableSheet } from '@/sheet-type';

describe('TableSheet Export Test', () => {
  it('should export correct data with series number', async () => {
    const s2 = new TableSheet(
      getContainer(),
      assembleDataCfg({
        meta: [
          {
            field: 'type',
            name: '产品类型',
            formatter: (type) => (type ? `${type}产品` : ''),
          },
        ],
        fields: {
          columns: ['province', 'city', 'type', 'sub_type', 'number'],
        },
      }),
      assembleOptions({
        seriesNumber: {
          enable: true,
        },
      }),
    );

    await s2.render();
    const data = await asyncGetAllPlainData({
      sheetInstance: s2,
      split: '\t',
      formatOptions: {
        formatHeader: true,
      },
    });

    const rows = data.split('\n');
    const headers = rows[0].split('\t');

    expect(rows).toHaveLength(78);
    expect(rows).toMatchSnapshot();

    // 6列数据 包括序列号
    rows.forEach((row) => {
      expect(row.split('\t')).toHaveLength(6);
    });

    expect(headers).toMatchSnapshot();
  });

  it('should export correct data with no series number', async () => {
    const s2 = new TableSheet(
      getContainer(),
      assembleDataCfg({
        meta: [],
        fields: {
          columns: ['province', 'city', 'type', 'sub_type', 'number'],
        },
      }),
      assembleOptions({
        seriesNumber: {
          enable: false,
        },
      }),
    );

    await s2.render();
    const data = await asyncGetAllPlainData({
      sheetInstance: s2,
      split: '\t',
    });
    const rows = data.split('\n');
    const headers = rows[0].split('\t');

    expect(rows).toHaveLength(78);
    expect(rows).toMatchSnapshot();

    // 5列数据 不包括序列号
    rows.forEach((e) => {
      expect(e.split('\t')).toHaveLength(5);
    });
    expect(headers).toMatchSnapshot();
  });
});

describe('PivotSheet Export Test', () => {
  it('should export correct data in grid mode', async () => {
    const s2 = new PivotSheet(
      getContainer(),
      assembleDataCfg({
        meta: [],
        fields: {
          valueInCols: true,
          columns: ['province', 'city', 'type', 'sub_type', 'number'],
        },
      }),
      assembleOptions({ hierarchyType: 'grid' }),
    );

    await s2.render();
    const data = await asyncGetAllPlainData({
      sheetInstance: s2,
      split: '\t',
    });
    const rows = data.split('\n');

    expect(rows).toHaveLength(14);
    expect(rows[0].split('\t')[1]).toEqual('province');
    expect(rows[1].split('\t')[1]).toEqual('city');

    rows.forEach((e) => {
      expect(e.split('\t')).toHaveLength(34);
    });
  });

  it('should export correct data in tree mode', async () => {
    const s2 = new PivotSheet(
      getContainer(),
      assembleDataCfg({
        meta: [],
        fields: {
          valueInCols: true,
          columns: ['province', 'city', 'type', 'sub_type', 'number'],
        },
      }),
      assembleOptions({
        hierarchyType: 'tree',
      }),
    );

    await s2.render();
    const data = await asyncGetAllPlainData({
      sheetInstance: s2,
      split: '\t',
    });
    const rows = data.split('\n');

    expect(rows).toHaveLength(16);
    expect(rows[0].split('\t')[1]).toEqual('province');
    expect(rows[1].split('\t')[1]).toEqual('city');

    rows.forEach((e) => {
      expect(e.split('\t')).toHaveLength(34);
    });
  });

  // 因为导出的数据单测，很难看出问题，所以提供图片 + 代码的模式查看：
  // https://gw.alipayobjects.com/zos/antfincdn/AU83KF1Sq/6fb3f3e6-0064-4ef8-a5c3-b1333fb59adf.png
  it('should export correct data in tree mode and row collapseAll is true', async () => {
    const s2 = new PivotSheet(
      getContainer(),
      assembleDataCfg(),
      assembleOptions({
        hierarchyType: 'tree',
        style: {
          rowCell: {
            collapseAll: true,
          },
        },
      }),
    );

    await s2.render();
    const data = await asyncGetAllPlainData({
      sheetInstance: s2,
      split: '\t',
    });
    const rows = data.split('\n');

    expect(rows).toHaveLength(5);
    expect(rows[0].split('\t').length).toEqual(5);
    expect(rows[0].split('\t')[0]).toEqual('类别');
    expect(rows[0].split('\t')[1]).toEqual('家具');
    expect(rows[1].split('\t')[0]).toEqual('子类别');
    expect(rows[1].split('\t')[1]).toEqual('桌子');
    expect(rows[2].split('\t')[0]).toEqual('省份');
    expect(rows[2].split('\t')[1]).toEqual('数量');
  });

  // https://gw.alipayobjects.com/zos/antfincdn/PyrWwocNf/56d0914b-159a-4293-8615-6c1308bf4b3a.png
  it('should export correct data in tree mode and hierarchyCollapse is false', async () => {
    const s2 = new PivotSheet(
      getContainer(),
      assembleDataCfg(),
      assembleOptions({
        hierarchyType: 'tree',
        style: {
          rowCell: {
            collapseAll: false,
          },
        },
      }),
    );

    await s2.render();
    const data = await asyncGetAllPlainData({
      sheetInstance: s2,
      split: '\t',
    });
    const rows = data.split('\n');

    expect(rows).toHaveLength(13);
    expect(rows[0].split('\t').length).toEqual(6);
    expect(rows[0].split('\t')[1]).toEqual('类别');
    expect(rows[0].split('\t')[2]).toEqual('家具');
    expect(rows[1].split('\t')[1]).toEqual('子类别');
    expect(rows[1].split('\t')[2]).toEqual('桌子');
    expect(rows[2].split('\t')[0]).toEqual('省份');
    expect(rows[2].split('\t')[1]).toEqual('城市');
    expect(rows[2].split('\t')[2]).toEqual('数量');
  });

  it('should export correct data in grid mode with valueInCols is false', async () => {
    const s2 = new PivotSheet(
      getContainer(),
      assembleDataCfg({
        meta: [],
        fields: {
          valueInCols: false,
          columns: ['province', 'city', 'type', 'sub_type', 'number'],
        },
      }),
      assembleOptions({
        hierarchyType: 'grid',
      }),
    );

    await s2.render();
    const data = await asyncGetAllPlainData({
      sheetInstance: s2,
      split: '\t',
    });
    const rows = data.split('\n');

    expect(rows).toHaveLength(13);
    rows.forEach((e) => {
      expect(e.split('\t')).toHaveLength(35);
    });
  });

  it('should export correct data in grid mode with totals in col', async () => {
    const s2 = new PivotSheet(
      getContainer(),
      assembleDataCfg({
        fields: {
          valueInCols: true,
          columns: ['province', 'city', 'type', 'sub_type', 'number'],
        },
      }),
      assembleOptions({
        hierarchyType: 'grid',
        totals: {
          row: {
            showGrandTotals: true,
            showSubTotals: true,
            subTotalsDimensions: ['province'],
          },
          col: {
            showGrandTotals: true,
            showSubTotals: true,
            subTotalsDimensions: ['type'],
          },
        },
      }),
    );

    await s2.render();
    const data = await asyncGetAllPlainData({
      sheetInstance: s2,
      split: '\t',
    });
    const rows = data.split('\n');

    expect(rows).toHaveLength(17);
    rows.forEach((e) => {
      expect(e.split('\t')).toHaveLength(53);
    });
  });

  it('should export correct data in grid mode with grouped totals in col', async () => {
    const s2 = new PivotSheet(
      getContainer(),
      assembleDataCfg({
        fields: {
          valueInCols: true,
          columns: ['province', 'city', 'type', 'sub_type', 'number'],
        },
      }),
      assembleOptions({
        hierarchyType: 'grid',
        totals: {
          row: {
            showGrandTotals: true,
            showSubTotals: true,
            subTotalsDimensions: ['province'],
          },
          col: {
            grandTotalsGroupDimensions: ['city', 'type'],
            subTotalsGroupDimensions: ['sub_type'],
            showGrandTotals: true,
            showSubTotals: true,
            subTotalsDimensions: ['type'],
          },
        },
      }),
    );

    await s2.render();
    const data = await asyncGetAllPlainData({
      sheetInstance: s2,
      split: '\t',
    });
    const rows = data.split('\n');

    expect(rows).toHaveLength(17);
    rows.forEach((e) => {
      expect(e.split('\t')).toHaveLength(60);
    });
  });

  it('should export correct data in grid mode with grouped totals in row', async () => {
    const s2 = new PivotSheet(
      getContainer(),
      assembleDataCfg({
        fields: {
          valueInCols: false,
          columns: ['province', 'city', 'type', 'sub_type', 'number'],
        },
      }),
      assembleOptions({
        hierarchyType: 'grid',
        totals: {
          row: {
            showGrandTotals: true,
            showSubTotals: true,
            subTotalsDimensions: ['province'],
          },
          col: {
            grandTotalsGroupDimensions: ['city', 'sub_type', 'province'],
            subTotalsGroupDimensions: ['sub_type'],
            showGrandTotals: true,
            showSubTotals: true,
            subTotalsDimensions: ['type'],
          },
        },
      }),
    );

    await s2.render();
    const data = await asyncGetAllPlainData({
      sheetInstance: s2,
      split: '\t',
    });
    const rows = data.split('\n');

    expect(rows).toHaveLength(16);
    rows.forEach((e) => {
      expect(e.split('\t')).toHaveLength(63);
    });
  });
  it('should export correct data in grid mode with totals in row', async () => {
    const s2 = new PivotSheet(
      getContainer(),
      assembleDataCfg({
        fields: {
          valueInCols: false,
          columns: ['province', 'city', 'type', 'sub_type', 'number'],
        },
      }),
      assembleOptions({
        hierarchyType: 'grid',
        totals: {
          row: {
            showGrandTotals: true,
            showSubTotals: true,
            subTotalsDimensions: ['province'],
          },
          col: {
            showGrandTotals: true,
            showSubTotals: true,
            subTotalsDimensions: ['type'],
          },
        },
      }),
    );

    await s2.render();
    const data = await asyncGetAllPlainData({
      sheetInstance: s2,
      split: '\t',
    });
    const rows = data.split('\n');

    expect(rows).toHaveLength(16);
    rows.forEach((e) => {
      expect(e.split('\t')).toHaveLength(54);
    });
  });

  it('should export correct data when isFormat: {formatHeader: true}', async () => {
    const s2 = new PivotSheet(
      getContainer(),
      assembleDataCfg({
        meta: [
          {
            field: 'province',
            formatter: (value) => {
              return `${value}-province`;
            },
          },
          {
            field: 'type',
            formatter: (value) => {
              return `${value}-type`;
            },
          },
        ],
        fields: {
          valueInCols: true,
          columns: ['province', 'city'],
          rows: ['type', 'sub_type'],
          values: ['number'],
        },
      }),
      assembleOptions({}),
    );

    await s2.render();
    const data = await asyncGetAllPlainData({
      sheetInstance: s2,
      split: '\t',
      formatOptions: {
        formatHeader: true,
      },
    });
    const rows = data.split('\n');

    expect(rows).toHaveLength(7);
    expect(rows[0].split('\t')[1]).toEqual('province');
    expect(rows[0].split('\t')[2]).toEqual('浙江省-province');
    expect(rows[1].split('\t')[1]).toEqual('city');
    expect(rows[3].split('\t')[0]).toEqual('家具-type');
  });

  it('should export correct $$extra$$ field name', async () => {
    const s2 = new PivotSheet(
      getContainer(),
      assembleDataCfg({
        meta: [
          {
            field: 'number',
            name: '数值',
          },
        ],
        fields: {
          valueInCols: true,
          columns: ['province', 'city'],
          rows: ['type', 'sub_type'],
          values: ['number'],
        },
      }),
      assembleOptions({}),
    );

    await s2.render();
    const data = await asyncGetAllPlainData({
      sheetInstance: s2,
      split: '\t',
      formatOptions: false,
    });
    const rows = data.split('\n');
    const headers = rows[2].split('\t');

    expect(headers).toMatchSnapshot();
  });

  describe('Custom Tree Export Test', () => {
    let s2: SpreadSheet;

    const getResult = async () => {
      const data = await asyncGetAllPlainData({
        sheetInstance: s2,
        split: '\t',
      });

      return data.split('\n');
    };

    beforeEach(async () => {
      const customRowDataCfg: S2DataConfig = {
        data: CustomGridData,
        meta: [
          {
            field: 'type',
            name: '类型',
          },
          {
            field: 'sub_type',
            name: '子类型',
          },
          {
            field: 'a-1',
            name: '层级1',
          },
          {
            field: 'a-2',
            name: '层级2',
          },
        ],
        fields: customRowGridSimpleFields,
      };

      s2 = new PivotSheet(getContainer(), customRowDataCfg, assembleOptions());

      await s2.render();
    });

    it('should export correct data in grid mode for custom row cell', async () => {
      s2.setOptions({ hierarchyType: 'grid' });
      await s2.render(false);

      const rows = await getResult();

      expect(rows).toMatchSnapshot();
    });

    it('should export correct data in tree mode for custom row cell', async () => {
      s2.setOptions({ hierarchyType: 'tree' });
      await s2.render(false);

      const rows = await getResult();

      expect(rows).toMatchSnapshot();
    });

    it('should export correct data in tree mode for custom row cell and custom corner text', async () => {
      s2.setOptions({ hierarchyType: 'tree', cornerText: '自定义' });
      await s2.render(false);

      const rows = await getResult();

      expect(rows).toMatchSnapshot();
    });
  });
});
