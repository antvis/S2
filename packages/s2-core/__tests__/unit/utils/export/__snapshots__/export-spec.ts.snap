// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PivotSheet Export Test Custom Tree Export Test should export correct data in grid mode for custom row cell 1`] = `
Array [
  "		类型	家具	家具
",
  "层级1	自定义节点 a-1-1	指标1	桌子	椅子
",
  "层级1	自定义节点 a-1-1	指标1	13	11
",
  "层级1	自定义节点 a-1-1	指标2	2	8
",
  "层级1	自定义节点 a-1-2			
",
  "层级2				",
]
`;

exports[`PivotSheet Export Test Custom Tree Export Test should export correct data in tree mode for custom row cell 1`] = `
Array [
  "		类型	家具	家具
",
  "		层级1/层级2/数值	桌子	椅子
",
  "层级1				
",
  "层级1	自定义节点 a-1-1			
",
  "层级1	自定义节点 a-1-1	指标1	13	11
",
  "层级1	自定义节点 a-1-1	指标2	2	8
",
  "层级1	自定义节点 a-1-2			
",
  "层级2				",
]
`;

exports[`PivotSheet Export Test Custom Tree Export Test should export correct data in tree mode for custom row cell and custom corner text 1`] = `
Array [
  "		类型	家具	家具
",
  "		自定义	桌子	椅子
",
  "层级1				
",
  "层级1	自定义节点 a-1-1			
",
  "层级1	自定义节点 a-1-1	指标1	13	11
",
  "层级1	自定义节点 a-1-1	指标2	2	8
",
  "层级1	自定义节点 a-1-2			
",
  "层级2				",
]
`;

exports[`PivotSheet Export Test should export correct $$extra$$ field name 1`] = `
Array [
  "type",
  "sub_type",
  "数值",
  "数值",
  "数值",
  "数值",
  "数值",
  "数值",
  "数值",
  "数值
",
]
`;

exports[`TableSheet Export Test should export correct data with no series number 1`] = `
Array [
  "province	city	type	sub_type	number
",
  "浙江省	杭州市	家具	桌子	7789
",
  "浙江省	绍兴市	家具	桌子	2367
",
  "浙江省	宁波市	家具	桌子	3877
",
  "浙江省	舟山市	家具	桌子	4342
",
  "浙江省	杭州市	家具	沙发	5343
",
  "浙江省	绍兴市	家具	沙发	632
",
  "浙江省	宁波市	家具	沙发	7234
",
  "浙江省	舟山市	家具	沙发	834
",
  "浙江省	杭州市	办公用品	笔	945
",
  "浙江省	绍兴市	办公用品	笔	1304
",
  "浙江省	宁波市	办公用品	笔	1145
",
  "浙江省	舟山市	办公用品	笔	1432
",
  "浙江省	杭州市	办公用品	纸张	1343
",
  "浙江省	绍兴市	办公用品	纸张	1354
",
  "浙江省	宁波市	办公用品	纸张	1523
",
  "浙江省	舟山市	办公用品	纸张	1634
",
  "四川省	成都市	家具	桌子	1723
",
  "四川省	绵阳市	家具	桌子	1822
",
  "四川省	南充市	家具	桌子	1943
",
  "四川省	乐山市	家具	桌子	2330
",
  "四川省	成都市	家具	沙发	2451
",
  "四川省	绵阳市	家具	沙发	2244
",
  "四川省	南充市	家具	沙发	2333
",
  "四川省	乐山市	家具	沙发	2445
",
  "四川省	成都市	办公用品	笔	2335
",
  "四川省	绵阳市	办公用品	笔	245
",
  "四川省	南充市	办公用品	笔	2457
",
  "四川省	乐山市	办公用品	笔	2458
",
  "四川省	成都市	办公用品	纸张	4004
",
  "四川省	绵阳市	办公用品	纸张	3077
",
  "四川省	南充市	办公用品	纸张	3551
",
  "四川省	乐山市	办公用品	纸张	352
",
  "		家具	桌子	26193
",
  "		家具		49709
",
  "		家具	沙发	23516
",
  "		办公用品		29159
",
  "		办公用品	笔	12321
",
  "		办公用品	纸张	16838
",
  "浙江省		家具	桌子	18375
",
  "浙江省		家具	沙发	14043
",
  "浙江省		办公用品	笔	4826
",
  "浙江省		办公用品	纸张	5854
",
  "四川省		家具	桌子	7818
",
  "四川省		家具	沙发	9473
",
  "四川省		办公用品	笔	7495
",
  "四川省		办公用品	纸张	10984
",
  "浙江省	杭州市	家具		13132
",
  "浙江省	杭州市	办公用品		2288
",
  "浙江省	杭州市			15420
",
  "浙江省	绍兴市	家具		2999
",
  "浙江省	绍兴市	办公用品		2658
",
  "浙江省	绍兴市			5657
",
  "浙江省	宁波市	家具		11111
",
  "浙江省	宁波市	办公用品		2668
",
  "浙江省	宁波市			13779
",
  "浙江省	舟山市	家具		5176
",
  "浙江省	舟山市	办公用品		3066
",
  "浙江省	舟山市			8242
",
  "四川省	成都市	家具		4174
",
  "四川省	成都市	办公用品		6339
",
  "四川省	成都市			10513
",
  "四川省	绵阳市	家具		4066
",
  "四川省	绵阳市	办公用品		3322
",
  "四川省	绵阳市			7388
",
  "四川省	南充市	家具		4276
",
  "四川省	南充市	办公用品		6008
",
  "四川省	南充市			10284
",
  "四川省	乐山市	家具		4775
",
  "四川省	乐山市	办公用品		2810
",
  "四川省	乐山市			7585
",
  "浙江省		家具		32418
",
  "浙江省		办公用品		10680
",
  "浙江省				43098
",
  "四川省		家具		17291
",
  "四川省		办公用品		18479
",
  "四川省				35770
",
  "				78868",
]
`;

exports[`TableSheet Export Test should export correct data with no series number 2`] = `
Array [
  "province",
  "city",
  "type",
  "sub_type",
  "number
",
]
`;

exports[`TableSheet Export Test should export correct data with series number 1`] = `
Array [
  "序号	province	city	产品类型	sub_type	number
",
  "1	浙江省	杭州市	家具	桌子	7789
",
  "2	浙江省	绍兴市	家具	桌子	2367
",
  "3	浙江省	宁波市	家具	桌子	3877
",
  "4	浙江省	舟山市	家具	桌子	4342
",
  "5	浙江省	杭州市	家具	沙发	5343
",
  "6	浙江省	绍兴市	家具	沙发	632
",
  "7	浙江省	宁波市	家具	沙发	7234
",
  "8	浙江省	舟山市	家具	沙发	834
",
  "9	浙江省	杭州市	办公用品	笔	945
",
  "10	浙江省	绍兴市	办公用品	笔	1304
",
  "11	浙江省	宁波市	办公用品	笔	1145
",
  "12	浙江省	舟山市	办公用品	笔	1432
",
  "13	浙江省	杭州市	办公用品	纸张	1343
",
  "14	浙江省	绍兴市	办公用品	纸张	1354
",
  "15	浙江省	宁波市	办公用品	纸张	1523
",
  "16	浙江省	舟山市	办公用品	纸张	1634
",
  "17	四川省	成都市	家具	桌子	1723
",
  "18	四川省	绵阳市	家具	桌子	1822
",
  "19	四川省	南充市	家具	桌子	1943
",
  "20	四川省	乐山市	家具	桌子	2330
",
  "21	四川省	成都市	家具	沙发	2451
",
  "22	四川省	绵阳市	家具	沙发	2244
",
  "23	四川省	南充市	家具	沙发	2333
",
  "24	四川省	乐山市	家具	沙发	2445
",
  "25	四川省	成都市	办公用品	笔	2335
",
  "26	四川省	绵阳市	办公用品	笔	245
",
  "27	四川省	南充市	办公用品	笔	2457
",
  "28	四川省	乐山市	办公用品	笔	2458
",
  "29	四川省	成都市	办公用品	纸张	4004
",
  "30	四川省	绵阳市	办公用品	纸张	3077
",
  "31	四川省	南充市	办公用品	纸张	3551
",
  "32	四川省	乐山市	办公用品	纸张	352
",
  "33			家具	桌子	26193
",
  "34			家具		49709
",
  "35			家具	沙发	23516
",
  "36			办公用品		29159
",
  "37			办公用品	笔	12321
",
  "38			办公用品	纸张	16838
",
  "39	浙江省		家具	桌子	18375
",
  "40	浙江省		家具	沙发	14043
",
  "41	浙江省		办公用品	笔	4826
",
  "42	浙江省		办公用品	纸张	5854
",
  "43	四川省		家具	桌子	7818
",
  "44	四川省		家具	沙发	9473
",
  "45	四川省		办公用品	笔	7495
",
  "46	四川省		办公用品	纸张	10984
",
  "47	浙江省	杭州市	家具		13132
",
  "48	浙江省	杭州市	办公用品		2288
",
  "49	浙江省	杭州市			15420
",
  "50	浙江省	绍兴市	家具		2999
",
  "51	浙江省	绍兴市	办公用品		2658
",
  "52	浙江省	绍兴市			5657
",
  "53	浙江省	宁波市	家具		11111
",
  "54	浙江省	宁波市	办公用品		2668
",
  "55	浙江省	宁波市			13779
",
  "56	浙江省	舟山市	家具		5176
",
  "57	浙江省	舟山市	办公用品		3066
",
  "58	浙江省	舟山市			8242
",
  "59	四川省	成都市	家具		4174
",
  "60	四川省	成都市	办公用品		6339
",
  "61	四川省	成都市			10513
",
  "62	四川省	绵阳市	家具		4066
",
  "63	四川省	绵阳市	办公用品		3322
",
  "64	四川省	绵阳市			7388
",
  "65	四川省	南充市	家具		4276
",
  "66	四川省	南充市	办公用品		6008
",
  "67	四川省	南充市			10284
",
  "68	四川省	乐山市	家具		4775
",
  "69	四川省	乐山市	办公用品		2810
",
  "70	四川省	乐山市			7585
",
  "71	浙江省		家具		32418
",
  "72	浙江省		办公用品		10680
",
  "73	浙江省				43098
",
  "74	四川省		家具		17291
",
  "75	四川省		办公用品		18479
",
  "76	四川省				35770
",
  "77					78868",
]
`;

exports[`TableSheet Export Test should export correct data with series number 2`] = `
Array [
  "序号",
  "province",
  "city",
  "产品类型",
  "sub_type",
  "number
",
]
`;
